shared_client.go

```diff
@@ -197,53 +337,75 @@ func handler(wg *sync.WaitGroup, client *Client, conn *Conn, requests chan reque
 		}
 	}()
 
-	type waiter struct {
-		ch    chan sharedClientResponse
-		start time.Time
-	}
-	waitingResponses := make(map[uint16]waiter)
+	waiters := newWaiterQueue()
 	defer func() {
 		conn.Close()
 		close(receiverTrigger)
 
-		// Drain responses send by receive loop to allow it to exit.
+		// Drain responses sent by receive loop to allow it to exit.
 		// It may be repeatedly reading after an i/o timeout, for example.
+		// This range will be done only after the receive loop has returned
+		// and closed 'responses' in its defer function.
 		for range responses {
 		}
 
-		for _, waiter := range waitingResponses {
-			waiter.ch <- sharedClientResponse{nil, 0, net.ErrClosed}
-			close(waiter.ch)
+		// Now cancel all the remaining waiters
+		waiters.FailAll(net.ErrClosed)
+
+		// Drain requests in case they come in while we are closing
+		// down. This loop is done only after 'requests' channel is closed in
+		// SharedClient.close() and it is not possible for new requests or timeouts
+		// to be sent on those closed channels.
+		for req := range requests {
+			req.ch <- sharedClientResponse{nil, 0, net.ErrClosed}
+			close(req.ch)
 		}
 	}()
 
+	deadline := time.NewTimer(0)
 	for {
+		// update timer
+		deadline.Reset(waiters.GetTimeout())
 		select {
+		case <-deadline.C:
+			waiters.Expired()
+
 		case req, ok := <-requests:
 			if !ok {
 				// 'requests' is closed when SharedClient is recycled, which happens
 				// responeses (or errors) have been received and there are no more
 				// requests to be sent.
 				return
 			}
-			start := time.Now()
 
 			// Check if we already have a request with the same id
 			// Due to birthday paradox and the fact that ID is uint16
 			// it's likely to happen with small number (~200) of concurrent requests
 			// which would result in goroutine leak as we would never close req.ch
-			if _, ok := waitingResponses[req.msg.Id]; ok {
-				req.ch <- sharedClientResponse{nil, 0, fmt.Errorf("duplicate request id %d", req.msg.Id)}
-				close(req.ch)
-				continue
+			if waiters.Exists(req.msg.Id) {
+				// find next available ID
+				duplicate := true
+				for id := req.msg.Id + 1; id != req.msg.Id; id++ {
```

Comment:
I'm not sure how safe it is to use kind of "predictable" Ids - https://nvd.nist.gov/vuln/detail/CVE-2008-0087 . I would probably try a few times (for example 3) `Id()` instead. That would also give a more predictable runtime.

Reply:
Each request starts from a random number (via `Id()`) by the caller), so it should not be likely that the Id's of consecutive requests would be sequential. This relies on the caller, though.

IMO 3 random tries might not work if there are a lot of requests in flight?

Reply:
> IMO 3 random tries might not work if there are a lot of requests in flight?

true, but that would require  ~10s of thousands of inflight requests - meaning probably something else is wrong already. I would consider using fixed-number as a circuit-breaker. 

For example, with 32k inflight requests, we would still have a chance of 1 - (32000/65000)^4 ~= 0.94 of getting a new correct ID.

Reply:
Done, with 5 tries.

Reply:
> I would consider using fixed-number as a circuit-breaker.

Did not get this, though?

Reply:
ah, nvm that's what I meant essentially, fixed loop.